File: kairos-web/src/types/diagnostic.ts
/**
 * Strict TypeScript Data Contracts for Kairos Quant Engine.
 */

export type HorizonMode = "COMPOUNDER" | "SWING";
export type MarketCapBucket = "LARGE_CAP" | "MID_CAP" | "SMALL_CAP";
export type PrimaryAction = "HOLD" | "TIGHTEN_STOP" | "TRIM_25" | "TRIM_50" | "EXIT_FULLY";

export type OverrideRule =
  | "NONE"
  | "TIER_1_HARD_GOVERNANCE_BYPASS"
  | "RULE_2A_STOP_BREACH_COMPOUNDER"
  | "RULE_2B_STOP_BREACH_SWING"
  | "RULE_1_COMPOUNDER_VOLATILITY_BUFFER"
  | "RULE_3_SELL_INTO_STRENGTH"
  | "RULE_4_DOUBLE_STRUCTURAL_BREAKDOWN"
  | "RULE_5_MOMENTUM_EXHAUSTION_DIVERGENCE";

export interface PlainLanguageExplanations {
  summary: string;
  pillar_fund: string;
  pillar_tech: string;
  pillar_quant: string;
  pillar_news: string;
}

export interface ScoreCard {
  s_fund: number;
  s_tech: number;
  s_quant: number;
  s_news: number;
  s_composite: number;
}

export interface PrecedenceWeights {
  w_fund: number;
  w_tech: number;
  w_quant: number;
  w_news: number;
  base_multiplier: number;
  net_multiplier: number;
}

export interface StopLossTelemetry {
  current_price: number;
  chandelier_stop: number;
  cushion_pct: number;
  atr_14: number;
  highest_high_22: number;
  is_stop_breached: boolean;
}

export interface RiskRewardTelemetry {
  target_price: number;
  reward_delta: number;
  risk_delta: number;
  risk_reward_ratio: number;
  quarter_kelly_pct: number;
}

export interface ChartDataPoint {
  time: number;
  open: number;
  high: number;
  low: number;
  close: number;
  stop: number;
}

export interface FundamentalMetrics {
  peg_ratio: number | null;
  roce_current: number;
  roce_3q_avg: number;
  promoter_pledge_pct: number;
  fcf_to_net_profit: number;
  debt_to_equity: number;
}

export interface TechnicalMetrics {
  sma_50: number;
  sma_200: number;
  rsi_14: number;
  delivery_pct: number;
}

export interface QuantMetrics {
  high_52w: number;
  realized_volatility_1y: number;
  beta: number;
}

export interface SentimentDisclosureInput {
  headline: string;
  hours_ago: number;
  sentiment_score: number;
  is_tier1_trigger: boolean;
}

export interface TaxImpactResult {
  shares_held: number;
  shares_to_sell: number;
  shares_retained: number;
  buy_price: number;
  current_price: number;
  gross_proceeds: number;
  capital_gain: number;
  tax_type: "STCG" | "LTCG";
  tax_rate_pct: number;
  tax_liability: number;
  net_cash_realized: number;
  new_breakeven_price: number;
  new_downside_cushion_pct: number;
}

export interface DiagnosticOutput {
  symbol: string;
  company_name: string;
  horizon_mode: HorizonMode;
  market_cap_bucket: MarketCapBucket;
  action: PrimaryAction;
  rule_applied: OverrideRule;
  explanation: string;
  scores: ScoreCard;
  weights: PrecedenceWeights;
  stop_telemetry: StopLossTelemetry;
  risk_telemetry: RiskRewardTelemetry;
  fundamentals: FundamentalMetrics;
  technicals: TechnicalMetrics;
  quant: QuantMetrics;
  disclosures: SentimentDisclosureInput[];
  chart_data: ChartDataPoint[];
  plain_language: PlainLanguageExplanations;
  audit_hash: string;
  evaluated_at_epoch: number;
  tax_impact?: TaxImpactResult | null;
}

export interface StockSearchResult {
  symbol: string;
  company_name: string;
  exchange: string;
  market_cap_bucket: MarketCapBucket;
  sector: string;
}

export type StageType =
  | "INITIALIZING"
  | "FETCHING_OHLCV"
  | "FETCHING_FUNDAMENTALS"
  | "SENTIMENT_ANALYSIS"
  | "RESOLVING_CONFLICTS"
  | "COMPLETE"
  | "ERROR";

export interface TelemetryEvent {
  stage: StageType;
  progress: number;
  message: string;
  data?: DiagnosticOutput | null;
}

File: kairos-engine/app/schemas/diagnostic.py
"""Diagnostic schemas and Pydantic data contracts for Kairos Quant."""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field, ConfigDict
from app.schemas.enums import (
    HorizonMode,
    MarketCapBucket,
    PrimaryAction,
    OverrideRule,
)


class PlainLanguageExplanations(BaseModel):
    """Contextual plain-language text representations for verdicts and pillars."""
    model_config = ConfigDict(frozen=True)
    
    summary: str
    pillar_fund: str
    pillar_tech: str
    pillar_quant: str
    pillar_news: str


class OHLCVBar(BaseModel):
    """Single OHLCV candlestick bar."""
    model_config = ConfigDict(frozen=True)
    
    time: int = Field(..., description="Unix epoch timestamp in seconds")
    open: float = Field(..., ge=0.0)
    high: float = Field(..., ge=0.0)
    low: float = Field(..., ge=0.0)
    close: float = Field(..., ge=0.0)
    volume: float = Field(..., ge=0.0)


class FundamentalMetricsInput(BaseModel):
    """Pillar A: Fundamental accounting and valuation inputs."""
    model_config = ConfigDict(frozen=True)
    
    peg_ratio: Optional[float] = Field(None, description="Price/Earnings to Growth ratio")
    roce_current: float = Field(..., description="Current Return on Capital Employed in %")
    roce_3q_avg: float = Field(..., description="Trailing 3-quarter average ROCE in %")
    promoter_pledge_pct: float = Field(0.0, ge=0.0, le=100.0, description="Promoter shares pledged in %")
    fcf_to_net_profit: float = Field(..., description="Free Cash Flow / Net Profit ratio")
    debt_to_equity: float = Field(0.0, ge=0.0, description="Total Debt / Total Equity ratio")


class TechnicalMetricsInput(BaseModel):
    """Pillar B: Technical moving averages and momentum inputs."""
    model_config = ConfigDict(frozen=True)
    
    sma_50: float = Field(..., ge=0.0)
    sma_200: float = Field(..., ge=0.0)
    rsi_14: float = Field(..., ge=0.0, le=100.0)
    delivery_pct: float = Field(..., ge=0.0, le=100.0, description="NSE Security Delivery %")


class QuantMetricsInput(BaseModel):
    """Pillar C: Volatility, drawdown, and beta inputs."""
    model_config = ConfigDict(frozen=True)
    
    high_52w: float = Field(..., ge=0.0)
    realized_volatility_1y: float = Field(..., ge=0.0, description="1-year annualized volatility in %")
    beta: float = Field(1.0, description="1-year Beta vs Nifty 50")


class SentimentDisclosureInput(BaseModel):
    """Pillar D: Structured corporate announcement input."""
    model_config = ConfigDict(frozen=True)
    
    headline: str
    hours_ago: float = Field(..., ge=0.0)
    sentiment_score: float = Field(..., ge=-1.0, le=1.0, description="FinBERT score: -1 to +1")
    is_tier1_trigger: bool = Field(False, description="Auditor resignation, SEBI probe, or debt default")


class ScoreCard(BaseModel):
    """Individual pillar scores and continuous composite score [0, 100]."""
    model_config = ConfigDict(frozen=True)
    
    s_fund: float = Field(..., ge=0.0, le=100.0)
    s_tech: float = Field(..., ge=0.0, le=100.0)
    s_quant: float = Field(..., ge=0.0, le=100.0)
    s_news: float = Field(..., ge=0.0, le=100.0)
    s_composite: float = Field(..., ge=0.0, le=100.0)


class PrecedenceWeights(BaseModel):
    """Resolved 2D Precedence Grid weights and multipliers."""
    model_config = ConfigDict(frozen=True)
    
    w_fund: float
    w_tech: float
    w_quant: float
    w_news: float
    base_multiplier: float
    net_multiplier: float


class StopLossTelemetry(BaseModel):
    """Chandelier ratcheting stop floor telemetry."""
    model_config = ConfigDict(frozen=True)
    
    current_price: float
    chandelier_stop: float
    cushion_pct: float
    atr_14: float
    highest_high_22: float
    is_stop_breached: bool


class RiskRewardTelemetry(BaseModel):
    """Fractional Kelly allocation and asymmetric risk-reward telemetry."""
    model_config = ConfigDict(frozen=True)
    
    target_price: float
    reward_delta: float
    risk_delta: float
    risk_reward_ratio: float
    quarter_kelly_pct: float


class ChartDataPoint(BaseModel):
    """Single point in time containing actual price and calculated stop-loss."""
    model_config = ConfigDict(frozen=True)
    
    time: int
    open: float
    high: float
    low: float
    close: float
    stop: float


class DiagnosticInput(BaseModel):
    """Canonical input payload required for full diagnostic evaluation."""
    model_config = ConfigDict(frozen=True)
    
    symbol: str
    company_name: str
    current_price: float = Field(..., ge=0.0)
    horizon_mode: HorizonMode
    market_cap_bucket: MarketCapBucket
    bars: List[OHLCVBar] = Field(..., min_length=22, description="At least 22 historical OHLCV bars")
    fundamentals: FundamentalMetricsInput
    technicals: TechnicalMetricsInput
    quant: QuantMetricsInput
    disclosures: List[SentimentDisclosureInput] = Field(default_factory=list)
    consensus_target_price: Optional[float] = None
    manual_atr_mult: Optional[float] = None


class DiagnosticOutput(BaseModel):
    """Canonical output contract produced by the Quant Math Engine."""
    model_config = ConfigDict(frozen=True)
    
    symbol: str
    company_name: str
    horizon_mode: HorizonMode
    market_cap_bucket: MarketCapBucket
    action: PrimaryAction
    rule_applied: OverrideRule
    explanation: str
    scores: ScoreCard
    weights: PrecedenceWeights
    stop_telemetry: StopLossTelemetry
    risk_telemetry: RiskRewardTelemetry
    fundamentals: FundamentalMetricsInput
    technicals: TechnicalMetricsInput
    quant: QuantMetricsInput
    disclosures: List[SentimentDisclosureInput]
    chart_data: List[ChartDataPoint]
    plain_language: PlainLanguageExplanations
    audit_hash: str
    evaluated_at_epoch: int
    tax_impact: Optional[Dict[str, Any]] = None

